numpy/core/src/npymath/npy_math_internal.h.src

```diff
@@ -845,3 +851,30 @@ npy_rshift@u@@c@(npy_@u@@type@ a, npy_@u@@type@ b)
 }
 /**end repeat1**/
 /**end repeat**/
+
+
+/**begin repeat
+ *
+ * #type  = ubyte, ushort, uint, ulong, ulonglong#
+ * #STYPE = BYTE,  SHORT,  INT,  LONG,  LONGLONG#
+ * #c     = ,      ,       ,     ll,    ll#
+ */
+#undef MAGIC_ARRAY
+#if 0
+/**begin repeat1
+ * #len = 8, 16, 32, 64#
+ */
+#elif NPY_BITSOF_@STYPE@ == @len@
+    #define MAGIC_ARRAY B@len@
+/**end repeat1**/
+#endif
+
+NPY_INPLACE npy_@type@
+npy_@type@_countbits(npy_@type@ a)
+{
+    a = a - ((a >> 1) & MAGIC_ARRAY[0]);
+    a = ((a & MAGIC_ARRAY[1])) + ((a >> 2) & MAGIC_ARRAY[1]);
+    return (((a + (a >> 4)) & MAGIC_ARRAY[2]) * MAGIC_ARRAY[3]) >> (NPY_BITSOF_@STYPE@ - 1)*8;
```

Comment:
Any suggestions on how to avoid a compiler warning on `(NPY_BITSOF_@STYPE@ - 1)*8`. I expected it to be computed at compile-time, but it actually throws a warning.

Reply:
What's the warning?

Reply:
Perhaps you could refer to a fast C/C++ bit population count library called [libpopcnt ](https://github.com/kimwalisch/libpopcnt).

Reply:
Thanks @Qiyu8 , the [library ](https://github.com/kimwalisch/libpopcnt/blob/master/libpopcnt.h)has a lot of useful stuff we can refer to.

Reply:
> What's the warning?

@eric-wieser ,  https://github.com/numpy/numpy/pull/19355/checks?check_run_id=2926152038#step:4:4113,
we are getting a `warning: right shift count >= width of type`

Reply:
I have fixed it in 975b8d1